### ISSUE TITLE
Cargo mutants command fails when using a local dependency

### ISSUE THREAD

**Issue by dmoka** [sourcefrog/cargo-mutants#42]
# Description

The `cargo mutants` command fails when there is a local dependency for a binary project.

This happens in a bigger project of us, but I created a test repo for the sake of easy reproduction.

## Reproduction

Steps:
1. Clone the following [repo](https://github.com/dmoka/test-repo-for-cargo-mutants)
2. Go to the `./main` directory
3. Execute `cargo mutants` command

Expected behavior:

- Some mutations are generated (if possible)

Actual behavior:

- The command fails with the following log:
```
Freshen source tree ... ok in 0.015s
Copy source and build products to scratch directory ... 14 MB in 0.006s
Unmutated baseline ... FAILED in 0.007s

*** baseline
*** run /home/dmoka/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/cargo check --tests
*** cargo result: Exited(101) in 0.007sency of package `main v0.1.0 (/tmp/cargo-mutants--eL1pHT.tmp)`

Caused by:
  failed to load source for dependency `utils`

Caused by:
  Unable to update /tmp/utils

Caused by:
  failed to read `/tmp/utils/Cargo.toml`

Caused by:
  No such file or directory (os error 2)
```

Note: when `cargo run` is executed in the `./main` directory, the binary is executed properly with the local `utils` dependency included. So I assume that it should have something to do with the cargo-mutants crate itself.

System: Ubunt 22.04

If you need any more information, let me know, thank you!

**Comment by sourcefrog** [comment]
Hi, thanks for the report and reproduction.

I think the problem here is that cargo-mutants makes a copy of the tree before mutating it, so any paths relative to the original location won't work, as in your reproduction tree.

https://github.com/sourcefrog/cargo-mutants#:~:text=dependencies%20by%20a%20relative%20path

Using an absolute path should work, but that's not always convenient, and I'd like for it to just work without the user needing to make any prior adjustments.

There seem to be a few options:

1. Rewrite `Cargo.toml` after copying to use absolute paths for any relative paths pointing out of the tree. Perhaps this is best. As a step towards this we could at least look for them and error cleanly.
2. Give an option to mutate the tree in place (and put it back afterwards.) This would avoid any patch related issues, but there's some risk that the mutation is actually not cleaned up properly if the program stops abruptly -- and anyhow the user would have to work out to use this option, which is not ideal.
3. An option to copy a larger, enclosing, directory, perhaps containing all the dependencies. But again the user needs to understand that they need it.

**Comment by sourcefrog** [comment]
- [ ] Find all `Cargo.toml` files in the tree to be copied: there might be some in subdirectories. (Possibly we should not copy nested crates and instead turn them into relative references?)
- [ ] Parse `Cargo.toml` probably using <https://docs.rs/toml/>
- [ ] Write out the possibly-modified file into the scratch directory.
- [ ] Walk the toml looking for `path` attributes, which might occur directly in the dependency or in an `patch` or `replace` section. (And, there are a couple of possible structures, but possibly the toml crate abstracts this.)
- [ ] For every relative path, replace it with an absolute path, resolved from the directory where it occurs.
- [ ] Also look for and modify `.cargo/config.toml`.

**Comment by sourcefrog** [comment]
https://github.com/sourcefrog/cargo-mutants/pull/43 fixes the particular case of `path` dependencies, although not yet all the other case. It does work on your reproduction repo. @dmoka , maybe you can try it on your bigger project?

**Comment by dmoka** [comment]
Wow, many thanks for your detailed investigation and your quick fix, much appreciated, Martin @sourcefrog !

Absolutely, happy to give it a try. I see that the latest crate version is still the `v0.2.7` which does not contain this latest fix, if I am not mistaken.

What is the easiest way to get the fixed version of the crate and trying out locally?

Thanks again, and have a great day!

**Comment by sourcefrog** [comment]
Yep, it's not released yet. I'm pretty sure this works but let's check it:

1. Clone this repo
2. Switch to the 'relpath' branch
3. `cargo install --path . `
4. Run `cargo mutants` on your tree

**Comment by dmoka** [comment]
Just checked it, and the relative-path issue has been indeed resolved.

Great job, thanks for the support!
